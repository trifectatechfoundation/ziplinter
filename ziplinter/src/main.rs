use rc_zip_sync::ReadZip;
use std::ops::Deref;

fn main() {
    let mut args = std::env::args();
    let _ = args.next();
    let path = args.next().unwrap();

    let file = std::fs::File::open(path).unwrap();
    let archive = file.read_zip().unwrap();

    eprintln!(
        "{}",
        serde_json::to_string_pretty(&archive.deref()).unwrap()
    )
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;
    use std::{error::Error, path::Path};

    fn process_zip_file(zip_path: &Path) -> Result<serde_json::Value, Box<dyn Error>> {
        let file = std::fs::File::open(zip_path).unwrap();
        let archive = file.read_zip();

        Ok(serde_json::to_value(archive?.deref())?)
    }

    #[test]
    fn snapshot_zip_files() {
        let fixtures_dir = std::env::current_dir()
            .unwrap()
            .join("../testdata")
            .canonicalize()
            .unwrap();

        println!("fixtures_dir: {}", fixtures_dir.display());

        for entry in std::fs::read_dir(fixtures_dir).expect("Failed to read fixtures directory") {
            let entry = entry.expect("Failed to read entry");
            let path = entry.path();

            if path
                .extension()
                .is_some_and(|e| e.eq_ignore_ascii_case("zip"))
            {
                let Ok(result) = process_zip_file(&path) else {
                    continue;
                };
                println!(
                    "current file: {}",
                    Path::new(path.file_name().unwrap()).display()
                );
                assert_json_snapshot!(
                    format!("{}", Path::new(path.file_name().unwrap()).display()),
                    result
                );
            }
        }
    }
}
