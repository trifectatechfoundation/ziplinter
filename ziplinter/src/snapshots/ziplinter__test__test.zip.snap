---
source: ziplinter/src/main.rs
expression: result
---
{
  "comment": "This is a zipfile comment.",
  "contents": [
    {
      "central": {
        "comment": "",
        "compressed_size": 25,
        "crc32": 3287144384,
        "creator_version": {
          "host_system": "Unix",
          "version": 30
        },
        "disk_nbr_start": 0,
        "external_attrs": 2175008768,
        "extra": [
          85,
          84,
          5,
          0,
          3,
          113,
          252,
          130,
          76,
          117,
          120,
          11,
          0,
          1,
          4,
          245,
          1,
          0,
          0,
          4,
          20,
          0,
          0,
          0
        ],
        "flags": 0,
        "header_offset": 0,
        "internal_attrs": 1,
        "method": "Deflate",
        "mode": 420,
        "modified": "2010-09-05T02:12:01Z",
        "name": "test.txt",
        "reader_version": {
          "host_system": "MsDos",
          "version": 20
        },
        "uncompressed_size": 26
      },
      "local": {
        "accessed": null,
        "compressed_size": 25,
        "crc32": 3287144384,
        "created": null,
        "extra": [
          85,
          84,
          9,
          0,
          3,
          113,
          252,
          130,
          76,
          118,
          252,
          130,
          76,
          117,
          120,
          11,
          0,
          1,
          4,
          245,
          1,
          0,
          0,
          4,
          20,
          0,
          0,
          0
        ],
        "flags": 0,
        "gid": 501,
        "header_offset": 0,
        "method": "Deflate",
        "method_specific": "None",
        "mode": 0,
        "modified": "2010-09-05T02:12:01Z",
        "name": "test.txt",
        "reader_version": {
          "host_system": "MsDos",
          "version": 20
        },
        "uid": 501,
        "uncompressed_size": 26
      }
    },
    {
      "central": {
        "comment": "",
        "compressed_size": 785,
        "crc32": 1423258110,
        "creator_version": {
          "host_system": "Unix",
          "version": 30
        },
        "disk_nbr_start": 0,
        "external_attrs": 2175008768,
        "extra": [
          85,
          84,
          5,
          0,
          3,
          58,
          48,
          131,
          76,
          117,
          120,
          11,
          0,
          1,
          4,
          245,
          1,
          0,
          0,
          4,
          20,
          0,
          0,
          0
        ],
        "flags": 0,
        "header_offset": 91,
        "internal_attrs": 0,
        "method": "Store",
        "mode": 420,
        "modified": "2010-09-05T05:52:58Z",
        "name": "gophercolor16x16.png",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uncompressed_size": 785
      },
      "local": {
        "accessed": null,
        "compressed_size": 785,
        "crc32": 1423258110,
        "created": null,
        "extra": [
          85,
          84,
          9,
          0,
          3,
          58,
          48,
          131,
          76,
          59,
          48,
          131,
          76,
          117,
          120,
          11,
          0,
          1,
          4,
          245,
          1,
          0,
          0,
          4,
          20,
          0,
          0,
          0
        ],
        "flags": 0,
        "gid": 501,
        "header_offset": 0,
        "method": "Store",
        "method_specific": "None",
        "mode": 0,
        "modified": "2010-09-05T05:52:58Z",
        "name": "gophercolor16x16.png",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uid": 501,
        "uncompressed_size": 785
      }
    }
  ],
  "encoding": "Utf8",
  "eocd": {
    "dir": {
      "inner": {
        "comment": [
          84,
          104,
          105,
          115,
          32,
          105,
          115,
          32,
          97,
          32,
          122,
          105,
          112,
          102,
          105,
          108,
          101,
          32,
          99,
          111,
          109,
          109,
          101,
          110,
          116,
          46
        ],
        "dir_disk_nbr": 0,
        "dir_records_this_disk": 2,
        "directory_offset": 954,
        "directory_records": 2,
        "directory_size": 168,
        "disk_nbr": 0
      },
      "offset": 1122
    },
    "dir64": null,
    "global_offset": 0
  },
  "parsed_ranges": [
    {
      "end": 1170,
      "kind": "eocd",
      "start": 1122
    },
    {
      "end": 1032,
      "kind": "cd header",
      "start": 954
    },
    {
      "end": 1122,
      "kind": "cd header",
      "start": 1032
    }
  ],
  "size": 1170
}
